Search

georgebrock opened the pull request:
Adds <kbd>/</kbd>, <kbd>?</kbd>, <kbd>n</kbd> and <kbd>N</kbd>.

Includes a refactor of arguments: Rather than operations knowing an argument type ("literal", "operation", "none") and using that to determine what to do with key presses, now there are argument objects (`LiteralArgument`, `CommandLineArgument`, `NullArgument` and `Operation` which has a compatible interface with arguments) and the operation just passed the key presses on for the argument to handle.

georgebrock committed:
Basic search implementation.

georgebrock committed:
Search doesn't always start on the current line.

georgebrock committed:
Search for nth match.

georgebrock committed:
Extract a method to search & move.

georgebrock committed:
Repeat searches with n and N.

georgebrock committed:
Add search commands to README.

georgebrock committed:
Search should wrap to start/end of document.

georgebrock committed:
Backward search.

georgebrock committed:
Improve descriptions of search commands.

georgebrock committed:
Add backward search to the README.

georgebrock committed:
Split Operation.keyPress into shorter methods.

georgebrock committed:
Command line args are just args now.

Breaks command line rendering, but makes the code much nicer.

georgebrock committed:
Use search commands as motions.

mike-burns commented on js/base.js:
```diff
@@ -258,26 +273,54 @@
+            return true;
+        } else {
+            return false;
+        }
```
This pattern!

... that's all.

georgebrock replied:
I've updated this:
- `moveCursor` now accepts an object, so I can pass the result of `findNext` directly to it
- `findNext` returns either `{row: …, col: …, found: true}` or `{found: false}`, never `null`

mike-burns commented:
Would a better name for `NullArgument` be `NoArgument`?

georgebrock committed:
Rename NullArgument to NoArgument.

georgebrock commented:
Good call, I've renamed it to `NoArgument`.

georgebrock committed:
More robust moveCursor method.

* Ignore null or undefined values.
* Accept an object with row and col properties.

georgebrock committed:
Refactor moveToNext and moveToLast.

Rather than explicitly handling null results, findNext and findLast now
return a useful object whether they find what we wanted or not.

georgebrock committed:
Use the more flexible moveCursor interface.

georgebrock commented:
Ready for another review, @mike-burns

mike-burns commented:
I like the refactoring. No further suggestions.

georgebrock commented:
Thanks Mike, merging.
